File: Data Structures & Algorithms/max-water-container/submission-2.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        
        int left = 0;
        int right = heights.size() - 1;

        int max_area = 0;
        while (left < right){
            int left_height = heights[left];
            int right_height = heights[right];

            int area = (right - left) * std::min(left_height, right_height);
            max_area = std::max(area, max_area);

            if (left_height <= right_height) {
                left++;
            } else {
                right--;
            }
        }
        return max_area;
    }
};

File: Data Structures & Algorithms/trapping-rain-water/submission-2.py
class Solution:
    def trap(self, height: List[int]) -> int:

        area = 0
        left, right = 0, len(height) - 1
        leftMax, rightMax = height[left], height[right]
        while left < right:
            if leftMax < rightMax:
                left += 1
                leftMax = max(leftMax, height[left])
                area += leftMax - height[left]
            else:
                right -= 1
                rightMax = max(rightMax, height[right])
                area += rightMax - height[right]
            
        return area


        